src/cpp/fastdds/domain/DomainParticipantImpl.hpp

```diff
@@ -337,7 +337,9 @@ class DomainParticipantImpl
     mutable std::mutex mtx_types_;
 
     //!Topic map
-    std::map<std::string, Topic*> topics_;
+    std::map<Topic*, TopicImpl*> topics_;
+    std::map<std::string, Topic*> topics_by_name_;
```

Comment:
Why do we need the Topic -> TopicImpl map? Each Topic contains a pointer to its implementation, so this map shouldn't be necessary.

Reply:
That was part of the leak problem.
- The Topic contains a pointer to the implementation
- The implementation contains a pointer to the Topic
- Topic::~Topic does nothing
- TopicImpl::~TopicImpl **deletes** the Topic pointer
I find this quite strange, but it is done like so also on subscriber and publisher.

So, we need the Topic->TopicImpl so that on the DomainParticipant destructor we can delete the TopicImpl and indirectly the Topic. (part of the leak was precisely that the TopicImpl and the Topic were not deleted).

Reply:
We can make the map a <string, TopicImpl>. Then we return impl->user_topic when necessary